From patchwork Wed Jul 25 17:51:35 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Peter Shier <pshier@google.com>
X-Patchwork-Id: 10544519
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id E5617157A
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed, 25 Jul 2018 17:51:50 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id D28972A980
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed, 25 Jul 2018 17:51:50 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id C6A0A2A994; Wed, 25 Jul 2018 17:51:50 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 651372A980
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed, 25 Jul 2018 17:51:50 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1729821AbeGYTEb (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Wed, 25 Jul 2018 15:04:31 -0400
Received: from mail-yw0-f202.google.com ([209.85.161.202]:40323 "EHLO
        mail-yw0-f202.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1728913AbeGYTEb (ORCPT <rfc822;kvm@vger.kernel.org>);
        Wed, 25 Jul 2018 15:04:31 -0400
Received: by mail-yw0-f202.google.com with SMTP id t10-v6so4718646ywc.7
        for <kvm@vger.kernel.org>; Wed, 25 Jul 2018 10:51:48 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=mime-version:date:message-id:subject:from:to:cc;
        bh=bnQhG1Sj33hgFKRbMKZjDNeHtLQCmXO2I2vmd1whWb0=;
        b=HXgcfM3niJQpoZ7dYCLi/2QHptEmXlLw34+Cb12CCtgQ6NVOZKHzKB8OoTUYJucm7a
         96Yx3HFdSd15129Nv1YjLxG+/Biv66e8JtaHCEpjsvqAdmEwkQ29vKUXDLS+pvyJU9+i
         +R9NS/AXpmuQ69dKE3efoA+UagHnTgpy1jrqG2Y6pfOs59orbu6qPD6LngWzITPxR93c
         1yI45lZa+gEaCHmfY04tYV27bwYIs6qmCozl+gMb6DStlCHaqRZPAnovqxwQvPAAn9V9
         vkuXAgpKQXofvgfOa1lIbwzn4ALqN00hVRhXcGAemwJlbkeak3dX7FMB0oGBV1rmbO7I
         8a0w==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:mime-version:date:message-id:subject:from:to:cc;
        bh=bnQhG1Sj33hgFKRbMKZjDNeHtLQCmXO2I2vmd1whWb0=;
        b=i954X41NYOlEhWbvCQ/FJFAozHKo1b+lf4RNeSPGX7Bz3eJiu1wwY/SHXFuMvvedxM
         D7aAfy8+9F3q+RH56/SwUEe28AztxyyUzbU/UW7ipk0rrccZXgFrr4vTHQYdRl5gBzf+
         AaQKuDB1Q7VV4Lfg48O37Ks8n9UtOTwfhqavzSPjLp2MOnJEOqnhtRNNwP7/WSOU+YPy
         vcKcKeEJztWJv0IErQ8iyt46wTa/6PIf1SiVQagk2K7jozbWAa3QDHjLHCE5jkEhjzNS
         1IrddL16fzSNV//H5nH9IHJnfErFl3tiLDFUzPAUDQ6PN7fxUtjXKHvoJsk/5NGlSdBl
         zrRw==
X-Gm-Message-State: AOUpUlG2h65J7fzbJzezJ6ZaJp5bYiDmrApBJxov96ftDDXHUBkWFuVH
        7oT24tQpNkzFfrhag2tT3ZLAL0xiRmvtvyOH2tU+a7TO4MRjbU6ko1y3FmmEwAsG3WxPlPMjQHc
        On6BeV1yZ+yiPS06iVrDrqpj+iK5dwCgnaJ4c+bsAlN+I+TcZb290MzmjHQ==
X-Google-Smtp-Source: 
 AAOMgpfOyA+qJaqGtGhc4nnJwGSmP1JTyY+0XxJ+8q0PzBY8+oj/bTVUCb6QyALf6gZc91BeVwkas7vBWzA=
MIME-Version: 1.0
X-Received: by 2002:a25:d652:: with SMTP id
 n79-v6mr6543937ybg.83.1532541107527;
 Wed, 25 Jul 2018 10:51:47 -0700 (PDT)
Date: Wed, 25 Jul 2018 10:51:35 -0700
Message-Id: <20180725175135.229236-1-pshier@google.com>
X-Mailer: git-send-email 2.18.0.233.g985f88cf7e-goog
Subject: [PATCH] Debugger extension: Add cr3 to debug/bp exception kvm_exit
 struct
From: Peter Shier <pshier@google.com>
To: kvm@vger.kernel.org
Cc: Ken Hofsass <hofsass@google.com>, Peter Shier <pshier@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

From: Ken Hofsass <hofsass@google.com>

Enables determining whether a debug/break event is associated with a guest
process of interest.

Signed-off-by: Ken Hofsass <hofsass@google.com>
Signed-off-by: Peter Shier <pshier@google.com>
Reviewed-by: Jim Mattson <jmattson@google.com>
---
 arch/x86/include/uapi/asm/kvm.h | 1 +
 arch/x86/kvm/vmx.c              | 2 ++
 arch/x86/kvm/x86.c              | 1 +
 include/uapi/linux/kvm.h        | 1 +
 4 files changed, 5 insertions(+)

diff --git a/arch/x86/include/uapi/asm/kvm.h b/arch/x86/include/uapi/asm/kvm.h
index c535c2fdea136..7aa39142a1b88 100644
--- a/arch/x86/include/uapi/asm/kvm.h
+++ b/arch/x86/include/uapi/asm/kvm.h
@@ -254,6 +254,7 @@ struct kvm_debug_exit_arch {
 	__u64 pc;
 	__u64 dr6;
 	__u64 dr7;
+	__u64 cr3;
 };
 
 #define KVM_GUESTDBG_USE_SW_BP		0x00010000
diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index 1689f433f3a08..a87ddef4c79bc 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -6751,6 +6751,7 @@ static int handle_exception(struct kvm_vcpu *vcpu)
 		rip = kvm_rip_read(vcpu);
 		kvm_run->debug.arch.pc = vmcs_readl(GUEST_CS_BASE) + rip;
 		kvm_run->debug.arch.exception = ex_no;
+		kvm_run->debug.arch.cr3 = kvm_read_cr3(vcpu);
 		break;
 	default:
 		kvm_run->exit_reason = KVM_EXIT_EXCEPTION;
@@ -6969,6 +6970,7 @@ static int handle_dr(struct kvm_vcpu *vcpu)
 			vcpu->run->debug.arch.dr6 = vcpu->arch.dr6;
 			vcpu->run->debug.arch.dr7 = dr7;
 			vcpu->run->debug.arch.pc = kvm_get_linear_rip(vcpu);
+			vcpu->run->debug.arch.cr3 = kvm_read_cr3(vcpu);
 			vcpu->run->debug.arch.exception = DB_VECTOR;
 			vcpu->run->exit_reason = KVM_EXIT_DEBUG;
 			return 0;
diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 0046aa70205aa..dd4d1ae37b794 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -2891,6 +2891,7 @@ int kvm_vm_ioctl_check_extension(struct kvm *kvm, long ext)
  	case KVM_CAP_SPLIT_IRQCHIP:
 	case KVM_CAP_IMMEDIATE_EXIT:
 	case KVM_CAP_GET_MSR_FEATURES:
+	case KVM_CAP_DEBUG_EVENT_PG_BASE_ADDR:
 		r = 1;
 		break;
 	case KVM_CAP_SYNC_REGS:
diff --git a/include/uapi/linux/kvm.h b/include/uapi/linux/kvm.h
index b6270a3b38e9f..1eab59739dcdc 100644
--- a/include/uapi/linux/kvm.h
+++ b/include/uapi/linux/kvm.h
@@ -949,6 +949,7 @@ struct kvm_ppc_resize_hpt {
 #define KVM_CAP_GET_MSR_FEATURES 153
 #define KVM_CAP_HYPERV_EVENTFD 154
 #define KVM_CAP_HYPERV_TLBFLUSH 155
+#define KVM_CAP_DEBUG_EVENT_PG_BASE_ADDR 156
 
 #ifdef KVM_CAP_IRQ_ROUTING
 
